seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/TaskExecutionService.java

```diff
@@ -18,134 +18,149 @@
-            this
-        );
-
-        executionContexts.put(task.getTaskID(), taskExecutionContext);
-        return taskExecutionContext;
+        CountDownLatch startedLatch = new CountDownLatch(tasks.size());
+        taskGroupExecutionTracker.blockingFutures = tasks
+            .stream()
+            .map(t -> new BlockingWorker(new TaskTracker(t, taskGroupExecutionTracker), startedLatch))
+            .map(executorService::submit)
+            .collect(toList());
 
+        // Do not return from this method until all workers have started. Otherwise
+        // on cancellation there is a race where the executor might not have started
+        // the worker yet. This would result in taskletDone() never being called for
+        // a worker.
+        uncheckRun(startedLatch::await);
     }
 
-    private final class TaskletTracker {
-        final NonCompletableFuture future = new NonCompletableFuture();
-        final Task task;
-        volatile Future<?> taskletFutures;
-
-        TaskletTracker(Task task, CompletableFuture<Void> cancellationFuture) {
-            this.task = task;
-
-            cancellationFuture.whenComplete(withTryCatch(logger, (r, e) -> {
-                if (e == null) {
-                    e = new IllegalStateException("cancellationFuture should be completed exceptionally");
-                }
-                future.internalCompleteExceptionally(e);
-                taskletFutures.cancel(true);
-            }));
-        }
-
-        @Override
-        public String toString() {
-            return "Tracking " + task;
+    public CompletableFuture<TaskExecutionState> submitTaskGroup(
+        TaskGroup taskGroup,
+        CompletableFuture<Void> cancellationFuture
+    ) {
+        Collection<Task> tasks = taskGroup.getTasks();
+        final TaskGroupExecutionTracker executionTracker = new TaskGroupExecutionTracker(cancellationFuture, taskGroup);
+        try {
+            ConcurrentMap<Long, TaskExecutionContext> taskExecutionContextMap = new ConcurrentHashMap<>();
+            final Map<Boolean, List<Task>> byCooperation =
+                tasks.stream()
+                    .peek(x -> {
+                        TaskExecutionContext taskExecutionContext = new TaskExecutionContext(x, nodeEngine);
+                        x.setTaskExecutionContext(taskExecutionContext);
+                        taskExecutionContextMap.put(x.getTaskID(), taskExecutionContext);
+                    })
+                    .collect(partitioningBy(Task::isThreadsShare));
+            submitThreadShareTask(executionTracker, byCooperation.get(true));
+            submitBlockingTask(executionTracker, byCooperation.get(false));
+            executionContexts.put(taskGroup.getId(), taskExecutionContextMap);
+        } catch (Throwable t) {
+            executionTracker.future.complete(new TaskExecutionState(taskGroup.getId(), ExecutionState.FAILED, t));
         }
+        return new NonCompletableFuture<>(executionTracker.future);
```

Comment:
I don't get why use `NonCompletableFuture`, if user want to cancel this job, how to do it?

Reply:
If the user wants to cancel, he needs to use the cancellationFuture.complete() parameter of this method.